### PR TITLE
Performance and code size improvements

### PR DESCRIPTION
**Pull request by Merri**
## Description
Here are the changes as discussed in the code review #38 including some further micro speed optimizations, code size suggestions and other generic improvements.

Code size has reduced:
- lib total: 30,692 to 28,193 bytes
- lib/es: 14,542 bytes to 13,272 bytes
- lib/js: 16,150 bytes to 14,921 bytes
- src total: 40,251 bytes to 37,734 bytes

So that equals to about 8 or 9% reduction in generated code size.

Also included is the idea to unobserve only once in each case.

## How has this been tested?
Each of the commits was run against the tests. I have not yet done any in-browser testing except with react-lazy which had largely similar changes to what has been done here.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](CONTRIBUTING.md) document
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: src/__tests__/observer.spec.js [hunk 1]
```diff
@@ -60,119 +54,119 @@ describe('#getPooled', () => {
...
-    test('observing a React instance when observer is not in storage yet', () => {
-        const observer = IntersectionObserverContainer.create(noop, defaultOptions);
+    test('observing a React instance when observer is not in observerElementsMap yet', () => {
+        const observer = createObserver(defaultOptions);
         const spy = jest.spyOn(observer, 'observe');
         const element = { target: { nodeType: 1, id: 2 }, observer };
-        IntersectionObserverContainer.observe(element);
-        const targets = Array.from(storage.get(observer));
+        observeElement(element);
+        const targets = Array.from(observerElementsMap.get(observer));
```
Comment (Rendez) [inline src/__tests__/observer.spec.js]: Should we use `observerElementsMap.get(observer).values().next().value` instead?

File: src/observer.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+export function getPooled(options = {}) {
+    const root = options.root || null;
+    const rootMargin = parseRootMargin(options.rootMargin);
+    const threshold = Array.isArray(options.threshold)
+        ? options.threshold
+        : [options.threshold != null ? options.threshold : 0];
+    const observers = observerElementsMap.keys();
+    let observer;
+    while ((observer = observers.next().value)) {
+        const unmatched =
```
Comment (Rendez) [inline src/observer.js]: Nice job reducing complexity and size

File: src/observer.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * The Intersection Observer API callback that is called whenever one element,
+ * called the target, intersects either the device viewport or a specified element.
+ * Also will get caled whenever the visibility of the target element changes and
+ * crosses desired amounts of intersection with the root.
+ * @param {array} changes
+ * @param {IntersectionObserver} observer
+ */
+export function callback(changes, observer) {
```
Comment (Rendez) [inline src/observer.js]: `forEach` was in fact smaller, what's the reason to change it for a `for` loop?

File: src/observer.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            element.handleChange(changes[i]);
+        }
+    }
+}
+
+export function createObserver(options) {
+    return getPooled(options) || new IntersectionObserver(callback, options);
+}
+
+export function observeElement(element) {
```
Comment (Rendez) [inline src/observer.js]: This was slightly smaller before, any reason for the change?

File: src/observer.js (hunk 3 above)
Comment (Merri) [reply src/observer.js]: Speed :) There have been improvements to `forEach` in JS engines, but it still seems to be a lot slower in tight loops. Of course overall this is a micro-level optimization. Only CPU would notice if it were aware of itself.

File: src/observer.js (hunk 4 above)
Comment (Merri) [reply src/observer.js]: Minified comparison:

```js
// before 134 bytes (ignoring additional class overhead):
var e=new Map;function observe(n){var v=void 0;if(e.has(n.v)){v=e.get(n.v)}else{v=new Set;e.set(n.v,v)}v.add(n);n.v.observe(n.target)}
// after 116 bytes:
var e=new Map;function observeElement(n){if(!e.has(n.o)){e.set(n.o,new Set)}e.get(n.o).add(n);n.o.observe(n.target)}
```

In addition I think the new version is a lot easier to understand, oftentimes with `let variable` and then filling it afterwards inside `if else` makes it harder to read what is happening.

File: src/__tests__/observer.spec.js (hunk 1 above)
Comment (Merri) [reply src/__tests__/observer.spec.js]: That would make sense in "real" code, in tests it only makes sense if it makes the test easier to understand.

**Pull request merged**